Framework/DataHandling/src/LoadILLSALSA.cpp

```diff
@@ -0,0 +1,186 @@
+  mustBePositive->setLower(0.0);
+  declareProperty("DetectorDistance", 1.0, mustBePositive, "Distance between the sample and the detector (meters)");
+  declareProperty("ThetaOffset", 0.0, "Offset for the 2theta value (degrees)");
+}
+
+/**
+ * Executes the algorithm
+ */
+void LoadILLSALSA::exec() {
+  const std::string filename = getPropertyValue("Filename");
+
+  // load scanned variables names
+  std::vector<std::string> scanVariableNames;
+  {
+    H5::H5File h5file(filename, H5F_ACC_RDONLY);
+    H5::Group scanVariables = h5file.openGroup("entry0/data_scan/scanned_variables/variables_names");
+    scanVariableNames = H5Util::readStringVector(scanVariables, "name");
+    scanVariables.close();
+    h5file.close();
+  }
+
+  // load data (detector and scanned variables)
+  Mantid::NeXus::NXRoot dataRoot(filename);
+  Mantid::NeXus::NXEntry dataFirstEntry = dataRoot.openFirstEntry();
+
+  Mantid::NeXus::NXData scanVariablesGroup = dataFirstEntry.openNXData("/data_scan/scanned_variables/data");
+  Mantid::NeXus::NXDouble scanVariables = scanVariablesGroup.openDoubleData();
+  scanVariables.load();
+  scanVariablesGroup.close();
+
+  Mantid::NeXus::NXData dataGroup = dataFirstEntry.openNXData("/data_scan/detector_data/data");
+  Mantid::NeXus::NXInt data = dataGroup.openIntData();
+  data.load();
+  dataGroup.close();
+
+  m_numberOfScans = data.dims(0);
+  m_numberOfRows = data.dims(1);
+  m_numberOfColumns = data.dims(2);
+
+  if (scanVariables.dims(1) != m_numberOfScans) {
+    std::ostringstream msg;
+    msg << "Number of scans in detector data (" << m_numberOfScans << ") ";
+    msg << "and scanned variables (" << scanVariables.dims(1) << ") do not match, please check your nexus file.";
+    throw std::runtime_error(msg.str());
+  }
+
+  m_outputWorkspace = DataObjects::create<DataObjects::Workspace2D>(m_numberOfRows * m_numberOfColumns + 1,
+                                                                    HistogramData::Points(m_numberOfScans));
+
+  // set the instrument
+  double sampleToDetectorDistance = getProperty("DetectorDistance");
+  Mantid::NeXus::NXFloat theta = dataFirstEntry.openNXFloat("/instrument/2theta/value");
+  theta.load();
+  double twoThetaAngle = theta[0] + static_cast<double>(getProperty("ThetaOffset"));
+  setInstrument(sampleToDetectorDistance, twoThetaAngle);
+
+  fillWorkspaceData(data, scanVariableNames, scanVariables);
+
+  dataRoot.close();
+
+  fillWorkspaceMetadata(filename);
+}
+
+void LoadILLSALSA::setInstrument(double distance, double angle) {
+  // load instrument
+  auto loadInst = createChildAlgorithm("LoadInstrument");
+  loadInst->setPropertyValue("InstrumentName", "SALSA");
+  loadInst->setProperty<API::MatrixWorkspace_sptr>("Workspace", m_outputWorkspace);
+  loadInst->setProperty("RewriteSpectraMap", Kernel::OptionalBool(true));
+  loadInst->execute();
+  setProperty("OutputWorkspace", m_outputWorkspace);
+
+  // translation
```

Comment:
Do you think it would make sense to extract placing of the instrument to own function?

Reply:
I set the instrument (loading + position) in a dedicated function: `LoadILLSALSA::setInstrument()`